strawberryfields/engine.py

```diff
@@ -270,8 +270,9 @@ def _broadcast_nones(val, dim):
                     p.reg_refs[k].val = v
 
             # if the program hasn't been compiled for this backend, do it now
-            if p.target != self.backend_name:
-                p = p.compile(self.backend_name, **compile_options) # TODO: shots might be relevant for compilation?
+            target = self.backend.circuit_spec
+            if target is not None and p.target != target:
+                p = p.compile(target, **compile_options)
```

Comment:
Do we want to make `Backend.circuit_spec` an abstract property that _must_ be overwritten? Or is there a use case for having backends with `Backend.circuit_spec = None`?

Reply:
`BaseBackend` :)

It's not an actual abc.ABC right now, so it can be instantiated.

Reply:
Oh, of course 🙂